exporters/otlp/src/otlp_http_client.cc

```diff
@@ -680,38 +777,130 @@ opentelemetry::sdk::common::ExportResult OtlpHttpClient::Export(
         << std::chrono::duration_cast<std::chrono::milliseconds>(options_.timeout).count()
         << " milliseconds)");
   }
-  bool write_successful = handler->waitForResponse();
 
-  // End the session
-  session->FinishSession();
+  // Wait for any session to finish if there are to many sessions
+  std::unique_lock<std::mutex> lock(session_waker_lock_);
+  bool wait_successful = session_waker_.wait_for(lock, options_.timeout, [this] {
+    std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+    return running_sessions_.size() <= options_.concurrent_sessions;
```

Comment:
What will happen if number of current running sessions are less than max-configured? `Export` will return success even though there is no real successful export ?

Reply:
Yes

Reply:
Yes, I can see the improvement we can achieve with these changes. The concern I have is that this deviates from the specs (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#interface-definition-1), which states that the Export method should return the result of transfer of the data over the wire. This can allow the Span Processor to take decision accordingly say retrying the failed transactions.
We may want to change the Export method to return the result callback to make these changes compliant to sepcs ?

Reply:
Also, had a quick look into how other SIGs are handling this. Java returns export status to processor asynchronously using [CompletableFuture](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html), while JS returns status using callback.

Reply:
FYI https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc-concurrent-requests

Reply:
> Also, had a quick look into how other SIGs are handling this. Java returns export status to processor asynchronously using [CompletableFuture](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html), while JS returns status using callback.

Good idea. Maybe we can use `nostd::variant<std::future<sdk::common::ExportResult>>` here. But it's a break change for exporters's APIs.

Reply:
> FYI https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc-concurrent-requests

The `OtlpGrpcExporter` and `OtlpGrpcLogExporter` have the same problem, but I didn't try to use asynchronize API or test the performance yet.

Reply:
> But it's a break change for exporters's APIs. I think it need more discussion about it.

Yes, I was thinking about that too. We have to keep supporting the existing `Exporter::Export()` interface and add new to support callback/futures. But it needs some thought and discussion.

Reply:
> The OtlpGrpcExporter and OtlpGrpcLogExporter have the same problem, but I didn't try to use asynchronize API or test the performance yet.

Agree, they too suffer from the same issue.